design-proposals/garbage-collect-completed-dvs.md

```diff
@@ -0,0 +1,60 @@
+# Overview
+DataVolumes are a useful construct for provisioning VM disks but once the PVC is provisioned and populated the DataVolume provides no value. This design presents an approach for garbage collecting completed DataVolumes. 
+
+## Motivation
+In the field, customers are confused by DVs after the fact. For example, when someone deletes a PVC that is owned by a DV they are confused when the PVC is re-created.
+
+## Goals
+DataVolumes will be garbage collected after their PVC is provisioned and populated.
+
+## Non Goals
+(limitations to the scope of the design)
+
+## Definition of Users
+(who is this feature set intended for)
+
+## User Stories
+* As a VM owner I use a standalone DataVolume to prepare a VM disk PVC for use by my VM. Once the DataVolume is complete, it has no use and can be confusing, so I want the system to garbage collect it. The PVC should still maintain an independent lifecycle from the VM.
+* As a VM owner I add an entry to the dataVolumeTemplates section of my VM spec to have a disk prepared for my VM. Once the DataVolume is complete, it has no use and can be confusing, so I want the system to garbage collect it. I want kubevirt to understand that it doesn't need to create a new DataVolume when the underlying PVC exists. The PVC should still share the same lifecycle as the VM.
+* As a VM owner I want to restore my backed up VM without needing to recreate the DataVolume.
+
+## Repos
+* **CDI**: Controllers (DataVolume/Import/Upload/Clone/DataImportCron/DataSource) will need adaptations:
+** So far the controllers assumed the existance of DV-PVC pairs, which is no longer a valid assumption.
+** DataImportCron controller currently re-creates the import DV if deleted, even if the PVC was already populated.
+* **KubeVirt**: VM controller adaptations
+# Design
+## PVC Ownership
+The DataVolume controller creates a PVC owned by its DV. When the PVC is populated:
+* PVC ownership by the DV will be removed
+* If DV has ownership (e.g. by VM, when created via DataVolumeTemplates), it will be transferred to its PVC
+* Any DV information that might be useful after it is garbage collected, should be annotated on the PVC (TBD)
+
+If the PVC population failed, the DV should remain the owner of the PVC.
+
+## Garbage Collection
+When DataVolume does not own a PVC, it can be deleted immediately, but we may want to keep it for a while:
+* Add CDI config minimal time for DV garbage collection after completion (disabled by default, for opt-in)
+  `dataVolumeTTL *int32`
+* Add DV annotation to prevent its garbage collection
+  `"cdi.kubevirt.io/storage.dv.retainAfterCompletion": "true"`
+
+Garbage collection will be executed whenever a DV phase is updated to Succeeded
+
+## API Examples
+(tangible API examples used for discussion)
+
+## Scalability
+(overview of how the design scales)
+
+## Update/Rollback Compatibility
```

Comment:
Am I correct in understanding that PVC Ownership and Garbage Collection are two separate processes?

Say we just upgraded KubeVirt/CDI to compatible versions that handle this spec.  Garbage collection is not enabled by default.

1.  Will new DataVolumes have PVC/DV ownership relation severed when DataVolume succeeded?
2.  Will existing succeeded DataVolumes have their ownership relations updated?

Now GC is enabled:

1.  Will all existing DataVolumes that are succeeded (and older than gc threshold) be deleted?

Reply:
It smells much cleaner both PVC/DV ownerRef update and GC are done together per single GC pass, so when GC is enabled all existing and future PVC/DV will be handled similarly.

Reply:
I agree that owner reference changes are an integral part of garbage collection and these two should always be handled together as one logical operation.

Reply:
@mhenriks  so you think it’s wiser that as first step instead of setting a global (CDI config) garbage collection policy, we may add specific per-DV annotations, e.g `deleteOnSuccess, recreatePvcOnDeletion, transferOwnerRefOnSuccess`?

Reply:
Although my questions about upgrade weren't answered directly, it seems that the answer were No, No, and Yes?

I am very weary of deleting DataVolumes that existed pre-upgrade when GC is enabled.  Maybe for DVs that are DataVolume templates it's okay though.  Otherwise, no.  They may have been created by users and own other resources.

Reply:
Sorry about my implicit answer, but you were right :)
I agree with your concerns. So instead of the purposed DV annotation `retainAfterCompletion` (default false) we may have explicit `deleteAfterCompletion` (default false), so new DVs can be created with it, and older pre-upgrade DVs won't be deleted?